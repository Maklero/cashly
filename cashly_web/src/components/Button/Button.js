import PropTypes from "prop-types";
import styled, { css } from "styled-components";

const buttonStyleSizes = {
    small: css`font-size: ${({ theme }) => theme.font.sizes.h6};`,
    medium: css`font-size: ${({ theme }) => theme.font.sizes.h5};`,
    large: css`font-size: ${({ theme }) => theme.font.sizes.h4};`
};

const buttonStyleVariants = {
    primary: css`
        color: ${({ theme }) => theme.colors.white};
        background-color: ${({ theme }) => theme.colors.primary400};

        &:hover { background-color: ${({ theme }) => theme.colors.primary500}; }
        
        &:active { background-color: ${({ theme }) => theme.colors.primary300}; }
    `,
    primaryOutline: css`
        color: ${({ theme }) => theme.colors.primary400};
        background-color: ${({ theme }) => theme.colors.white};
        border: 2px solid ${({ theme }) => theme.colors.primary400};

        &:hover { 
            color: ${({ theme }) => theme.colors.white};
            background-color: ${({ theme }) => theme.colors.primary500}; 
        }

        &:active {
            color: ${({ theme }) => theme.colors.white};
            background-color: ${({ theme }) => theme.colors.primary300};
        }
    `
};

const StyledButton = styled.button`
    border: none;
    outline: none;
    display: block;
    cursor: pointer;
    transition: .15s;
    border-radius: 2px;
    display: inline-block;
    padding: .5rem 1.5rem .5rem 1.5rem;
    font-family: ${({ theme }) => theme.font.family};
    font-weight: ${({ theme }) => theme.font.weights.semiBold};

    
    ${({ size }) => buttonStyleSizes[size]};
    ${({ variant }) => buttonStyleVariants[variant]};

    ${({ fullWidth }) => fullWidth && css`width: 100%;`};
    ${({ as }) => as === "a" && css`text-align: center; text-decoration: none;`};
`;

function Button(props) {
    const { children, variant, size, as } = props;

    return (
        <StyledButton
            as={as}
            size={size} 
            variant={variant}
            {...props}
        >
            {children}
        </StyledButton>
    );
}

Button.defaultProps = {
    as: "button",
    size: "medium",
    fullWidth: false,
    variant: "primary" 
};

Button.propTypes = {
    fullWidth: PropTypes.bool,
    children: PropTypes.node.isRequired,
    as: PropTypes.oneOf(["button", "a"]),
    size: PropTypes.oneOf(["small", "medium", "large"]),
    variant: PropTypes.oneOf(["primary", "primaryOutline"])
};

export default Button;