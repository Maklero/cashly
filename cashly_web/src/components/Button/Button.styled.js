import styled, { css } from "styled-components";

const BaseButton = styled.button`
  border: none;
  outline: none;
  cursor: pointer;
  line-height: 1.5;
  padding: 6px 16px;
  position: relative;
  border-radius: 2px;
  align-items: center;
  display: inline-flex;
  min-width: max-content;
  justify-content: center;
  font-size: ${({ theme }) => theme.fontSizes.h5};
  font-family: ${({ theme }) => theme.fontFamily};
  transition: ${({ theme }) => theme.defaultTransition};
  font-weight: ${({ theme }) => theme.fontWeights.regular};

  &:hover {
    background-color: ${({ theme }) => theme.colors.primary500};
  }
  &:active {
    background-color: ${({ theme }) => theme.colors.primary300};
  }

  ${({ fullWidth }) =>
    fullWidth &&
    css`
      width: 100%;
    `};
`;

const PrimaryButton = styled(BaseButton)`
  color: white;
  background-color: ${({ theme }) => theme.colors.primary400};
`;

const PrimaryOutlinedButton = styled(BaseButton)`
  padding: 5px 15px;
  background-color: transparent;
  border: 1px solid ${({ theme }) => theme.colors.primary400};

  &:hover {
    color: white;
  }
`;

const TextButton = styled(BaseButton)`
  background-color: transparent;
  color: ${({ theme }) => theme.colors.primary400};

  &:hover {
    background-color: ${({ theme }) => theme.colors.gray100};
  }

  &:active {
    background-color: ${({ theme }) => theme.colors.gray300};
  }
`;

const BaseIconWrapper = styled.span`
  display: inherit;
`;

const StartIconWrapper = styled(BaseIconWrapper)`
  margin-right: 16px;
`;

const EndIconWrapper = styled(BaseIconWrapper)`
  margin-left: 16px;
`;

export {
  PrimaryButton,
  PrimaryOutlinedButton,
  TextButton,
  StartIconWrapper,
  EndIconWrapper,
};
