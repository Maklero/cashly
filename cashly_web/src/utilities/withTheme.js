import { ThemeProvider } from "styled-components";
import { theme } from "../styles";

const withTheme = (Component) => {
    return (
        <ThemeProvider theme={theme}>
            <Component />
        </ThemeProvider>
    );
}

export default withTheme;