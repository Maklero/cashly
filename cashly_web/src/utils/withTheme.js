import { ThemeProvider } from 'styled-components'

import { theme } from '@/styles'

export const withTheme = (component) => (
    <ThemeProvider theme={theme}>{component}</ThemeProvider>
)
