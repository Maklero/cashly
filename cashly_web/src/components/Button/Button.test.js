import { render, screen } from "@testing-library/react";
import Button from "./Button";
import withTheme from "../../utilities/withTheme";

test("renders default variant of button component", () => {
    const PreparedButtonComponent = () => {
        return <Button>Hello, World!</Button>;
    };
    
    render(withTheme(PreparedButtonComponent));

    const buttonElm = screen.getByText("Hello, World!");

    expect(buttonElm).toMatchSnapshot();
    expect(buttonElm).toBeInTheDocument();
})

test("renders primary variant of button component", () => {
    const PreparedButtonComponent = () => {
        return <Button variant="primary">Hello, World!</Button>;
    }
    
    render(withTheme(PreparedButtonComponent));

    const buttonElm = screen.getByText("Hello, World!");

    expect(buttonElm).toBeInTheDocument();
})

test("renders primaryOutline variant of button component", () => {
    const PreparedButtonComponent = () => {
        return <Button variant="primaryOutline">Hello, World!</Button>;
    }
    
    render(withTheme(PreparedButtonComponent));

    const buttonElm = screen.getByText("Hello, World!");

    expect(buttonElm).toBeInTheDocument();
})

test("renders small size of button component", () => {
    const PreparedButtonComponent = () => {
        return <Button size="small">Hello, World!</Button>;
    }
    
    render(withTheme(PreparedButtonComponent));

    const buttonElm = screen.getByText("Hello, World!");

    expect(buttonElm).toBeInTheDocument();
})

test("renders medium size of button component", () => {
    const PreparedButtonComponent = () => {
        return <Button size="medium">Hello, World!</Button>;
    }
    
    render(withTheme(PreparedButtonComponent));

    const buttonElm = screen.getByText("Hello, World!");

    expect(buttonElm).toBeInTheDocument();
})

test("renders large size of button component", () => {
    const PreparedButtonComponent = () => {
        return <Button size="large">Hello, World!</Button>;
    }
    
    render(withTheme(PreparedButtonComponent));

    const buttonElm = screen.getByText("Hello, World!");

    expect(buttonElm).toBeInTheDocument();
    expect(buttonElm.nodeName).toBe("BUTTON");
})

test("renders button component as link", () => {
    const PreparedButtonComponent = () => {
        return <Button as="a">Hello, World!</Button>;
    }
    
    render(withTheme(PreparedButtonComponent));

    const buttonElm = screen.getByText("Hello, World!");

    expect(buttonElm).toMatchSnapshot();
    expect(buttonElm).toBeInTheDocument();
})