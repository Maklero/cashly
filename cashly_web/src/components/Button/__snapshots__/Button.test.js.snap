// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders button component as link 1`] = `
<a
  class="sc-bdvvtL fscJpZ"
>
  Hello, World!
</a>
`;

exports[`renders default variant of button component 1`] = `
<button
  class="sc-bdvvtL loGXrD"
>
  Hello, World!
</button>
`;
